File: Peak Index in a Mountain Array/Peak_Index_in_a_Mountain_Array.py
# Runtime: 32 ms, faster than 100.00% of Python3 online submissions for Peak Index in a Mountain Array.
# Memory Usage: 13.5 MB, less than 6.90% of Python3 online submissions for Peak Index in a Mountain Array.

class Solution:
    def peakIndexInMountainArray(self, A: 'List[int]') -> 'int':
        
        leftPtr, rightPtr = 0, len(A) - 1
        while leftPtr < rightPtr:
            midPtr = (leftPtr + rightPtr) // 2

            if A[midPtr - 1] < A[midPtr] < A[midPtr + 1]:
                leftPtr = midPtr + 1
            elif A[midPtr - 1] > A[midPtr] > A[midPtr + 1]:
                rightPtr = midPtr - 1
            else:
                leftPtr = midPtr
                break
        
        if A[leftPtr - 1] < A[leftPtr] > A[leftPtr + 1]:
            return leftPtr
        else:
            return rightPtr

File: Peak Index in a Mountain Array/Peak_Index_in_a_Mountain_Array.cpp
// Runtime: 16 ms, faster than 97.68% of C++ online submissions for Peak Index in a Mountain Array.
// Memory Usage: 9.6 MB, less than 13.82% of C++ online submissions for Peak Index in a Mountain Array.

class Solution 
{
public:
    int peakIndexInMountainArray(vector<int>& A) 
    {
        int leftPtr = 0, rightPtr = A.size() - 1;
        while (leftPtr < rightPtr)
        {
            int midPtr = (leftPtr + rightPtr) / 2;
            
            // 上升序列
            if (A[midPtr] < A[midPtr + 1] && A[midPtr] > A[midPtr - 1])
                leftPtr = midPtr + 1;
            // 下降序列
            else if (A[midPtr] > A[midPtr + 1] && A[midPtr] < A[midPtr - 1])
                rightPtr = midPtr - 1;
            // 刚好是峰值
            else
            {
                leftPtr = midPtr;
                break;
            }
        }
        
        if (A[leftPtr] > A[leftPtr + 1] && A[leftPtr] > A[leftPtr - 1])
            return leftPtr;
        else
            return rightPtr;
    }
};